### PR TITLE
[dashboard] [experimental] Add py-spy profiler support to Node page view

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This adds basic support for profiling via py-spy via the dashboard Nodes page. Two new links are added to the node > worker view: "Stack Trace", and "Flame Graph":

graphs

If you click "Stack Trace", this opens a new tab with the native stack:

graphs

If you click "Flame Graph", this returns a flame graph in a new tab after 5 seconds:

graphs

This functionality requires py-spy to have root privileges, Ray to have root privileges, or passwordless sudo to be available. If none of these is true, you get an error:

graphs

- [x] TODO: test on a multinode cluster

### CODE REVIEW

File: dashboard/modules/reporter/profile_manager.py
Comment (rkooo567) [inline dashboard/modules/reporter/profile_manager.py:74]: What do you think about making native optional? I believe most of the users won't need it (and it is noise) except some core devs like us?

File: dashboard/modules/reporter/profile_manager.py
Comment (rkooo567) [inline dashboard/modules/reporter/profile_manager.py:48]: same as here

File: dashboard/modules/reporter/reporter_head.py
Comment (rkooo567) [inline dashboard/modules/reporter/reporter_head.py:188]: Remove it?

File: dashboard/modules/reporter/reporter_head.py
Comment (rkooo567) [inline dashboard/modules/reporter/reporter_head.py:191]: Do you know what's the max output reply size for aiohttp? Maybe we can log warning if the size is too big?

File: dashboard/modules/reporter/reporter_head.py
Comment (rkooo567) [inline dashboard/modules/reporter/reporter_head.py:168]: Remove this too!

**Review by rkooo567 — COMMENTED** [review]
Mostly LGTM. Do you plan to have a CLI command for the same feature (maybe in a follow up)? I think that'll be very convenient. Also should we allow users to set duration from UI? (can also be in the future PR)

File: dashboard/client/src/pages/node/NodeRow.tsx
Comment (alanwguo) [inline dashboard/client/src/pages/node/NodeRow.tsx:213]: As a follow-up, we should probably add some sort of hint or tooltip that says that this will begin profiling for the next 5 seconds after being clicked.

File: python/ray/tests/test_dashboard_profiler.py
Comment (alanwguo) [inline python/ray/tests/test_dashboard_profiler.py:21]: I don't see any code to add py-spy to any requirements.
`py-spy` is already installed in the test machines?

**Review by alanwguo — APPROVED** [review]
UI changes lgtm

File: dashboard/modules/reporter/profile_manager.py
Comment (ericl) [reply dashboard/modules/reporter/profile_manager.py:74]: I find native very useful when using libraries like pytorch and such as well. I think we can leave it on and see if it's a burden in practice. It's better to avoid options when possible.

File: dashboard/modules/reporter/reporter_head.py
Comment (ericl) [reply dashboard/modules/reporter/reporter_head.py:191]: Added some logging improvements for these two handlers.

File: python/ray/_raylet.pyx
Comment (ericl) [inline python/ray/_raylet.pyx:765]: The parens was triggering a bug in py-spy, unfortunately.

File: python/ray/tests/test_dashboard_profiler.py
Comment (ericl) [reply python/ray/tests/test_dashboard_profiler.py:21]: Yeah, it's in our requirements.txt, which should be installed on CI (let's see).

File: dashboard/client/src/pages/node/NodeRow.tsx
Comment (ericl) [reply dashboard/client/src/pages/node/NodeRow.tsx:217]: Yeah--- I was thinking 5 seconds was short enough to be ok, but if we allow configuring longer durations then a dialog confirm would be useful.

**Comment by ericl** [comment]
> Mostly LGTM. Do you plan to have a CLI command for the same feature (maybe in a follow up)? I think that'll be very convenient. Also should we allow users to set duration from UI? (can also be in the future PR)

I'd like to keep focus on the dashboard for now. Many users report they have difficulty using the CLI (e.g., using ray client, or otherwise lacking shell access). The dashboard is the least common denominator and easiest to use tool.

Duration--- you can set it via the URL param today and should be easy to upgrade to a dialog later on.

**Comment by ericl** [comment]
Tested this e2e on a multi-node cluster and it seems to work fine.

**Comment by rkooo567** [comment]
@ericl sounds good for native + CLI. I think we can also do some dogfooding with ML team after merge the PR for more feedback!

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
Added @richardliaw for the doc reviewer.

**Review by richardliaw — APPROVED** [review]
trusting that this will work on anyscale at some point soon

**Comment by ericl** [comment]
Seems the issue is we don't have CAP_SYS_PTRACE in our CI. @simon-mo @rkooo567 is this something possible to add?

**Comment by simon-mo** [comment]
Yes. You can add add-caps argument to this file <https://github.com/ray-
project/buildkite-ci-pipelines/blob/main/ray_ci/step.json>

Reference <https://github.com/buildkite-plugins/docker-buildkite-plugin>

Once its pushed to the master new builds should pick it up. 

On November 5, 2022, GitHub ***@***.***> wrote:
> Seems the issue is we don't have CAP_SYS_PTRACE in our CI. @simon-mo
> <https://github.com/simon-mo>
> @rkooo567 <https://github.com/rkooo567> is this something possible to
> add?
>

—
Reply to this email directly, view it on GitHub <https://github.com/ray-
project/ray/pull/30011#issuecomment-1304602690>, or unsubscribe
<https://github.com/notifications/unsubscribe-
auth/AFBD7A537GN2NCDE4WKYHYLWG2OZTANCNFSM6AAAAAARWYZK2M>.
You are receiving this because you were mentioned.Message ID: <ray-
***@***.***>

**Comment by ericl** [comment]
> @ericl can you also update the doc / tooltip / or error message to explain this? Maybe in this PR, it could be just a part of error msg

Sounds good. Specifically, you mean the tooltip for the links?

**Comment by ericl** [comment]
Hmm, I'm still seeing the PTRACE error. @simon-mo any idea what's going on?

```

=== stderr ===
  | Permission Denied
  |  
  | It looks like you are running in a docker container. Please make sure you started your container with the SYS_PTRACE capability. See https://github.com/benfred/py-spy#how-do-i-run-py-spy-in-docker for more details
 

```

**Comment by simon-mo** [comment]
Let's try this... https://github.com/ray-project/buildkite-ci-pipelines/pull/8

**Pull request merged**
